pkg/client/cache/watch_cache.go

```diff
@@ -225,9 +225,11 @@ func (w *WatchCache) SetOnEvent(onEvent func(watch.Event)) {
 	w.onEvent = onEvent
 }
 
-func (w *WatchCache) GetAllEventsSince(resourceVersion uint64) ([]watch.Event, error) {
+// This function locks the whole cache for and doesn't release the lock on return.
+// The caller should explicitly unlock the cache by calling the returned function.
+func (w *WatchCache) GetAllEventsSinceAndLock(resourceVersion uint64) ([]watch.Event, func(), error) {
 	w.RLock()
-	defer w.RUnlock()
+	unlockFunc := w.RUnlock
```

Comment:
can't you just make this function thread unsafe and mandate the caller takes the rlock? I think distributed locking will be a pain to debug, especially in the watch event driven critical path of a high qps server.